### PR TITLE
promise_jsoo: init at 0.3.1

### PR DESCRIPTION
**Pull request by bhootd**
###### Description of changes

Js_of_ocaml bindings to JS Promises with supplemental functions 

https://github.com/mnxn/promise_jsoo

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{ lib, fetchurl, buildDunePackage, ocaml, js_of_ocaml, ppxlib, js_of_ocaml-ppx, gen_js_api, ojs }:
+
+let
+  owner = "mnxn";
+in
+buildDunePackage rec {
+  pname = "promise_jsoo";
+  version = "0.3.1";
+  minimumOCamlVersion = "4.02";
+  useDune2 = lib.versionAtLeast ocaml.version "4.08";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/promise_jsoo/default.nix:10]:
```suggestion
```

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    sha256 = "00pjnsbv0yv3hhxbbl8dsljgr95kjgi9w8j1x46gjyxg9zayrxzl";
+  };
+
+  buildInputs = [
+    js_of_ocaml
+    ppxlib
+    js_of_ocaml-ppx
+    gen_js_api
+  ];
+  propagatedBuildInputs = [ ojs ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/promise_jsoo/default.nix:23]: Are you sure that `js_of_ocaml` need not be also propagated?

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  meta = {
+    homepage = "https://github.com/${owner}/${pname}";
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/promise_jsoo/default.nix:26]:
```suggestion
    homepage = "https://github.com/mnxn/promise_jsoo";
```

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    description = "Js_of_ocaml bindings to JS Promises with supplemental functions";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ jayesh-bhoot ];
+  };
+}
+
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/promise_jsoo/default.nix:32]: There is one new line to much here.

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  owner = "mnxn";
+in
+buildDunePackage rec {
+  pname = "promise_jsoo";
+  version = "0.3.1";
+  minimumOCamlVersion = "4.02";
+  useDune2 = lib.versionAtLeast ocaml.version "4.08";
+
+  src = fetchurl {
+    url = "https://github.com/${owner}/${pname}/releases/download/v${version}/${pname}-v${version}.tbz";
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/promise_jsoo/default.nix:13]:
```suggestion
    url = "https://github.com/mnxn/promise_jsoo/releases/download/v${version}/promise_jsoo-v${version}.tbz";
```

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
+{ lib, fetchurl, buildDunePackage, ocaml, js_of_ocaml, ppxlib, js_of_ocaml-ppx, gen_js_api, ojs }:
+
+let
+  owner = "mnxn";
+in
+buildDunePackage rec {
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/promise_jsoo/default.nix:6]:
```suggestion
buildDunePackage rec {
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 2 above)
Comment (jayesh-bhoot) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix]: Actually, `ojs` too doesn't need to be required. I made a mistake in my flake's buildInputs.

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 1 above)
Comment (jayesh-bhoot) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix]: Outdated due to newer commits

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix [hunk 7]
```diff
@@ -0,0 +1,27 @@
+{ lib, fetchurl, buildDunePackage, ocaml, js_of_ocaml, ppxlib, js_of_ocaml-ppx, gen_js_api }:
+
+buildDunePackage rec {
+  pname = "promise_jsoo";
+  version = "0.3.1";
+  minimumOCamlVersion = "4.08";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/promise_jsoo/default.nix:6]:
```suggestion
  minimalOCamlVersion = "4.08";
```

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 2 above)
Comment (vbgl) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix]: I think that both `ojs` and `js_of_ocaml` must be added to `propagatedBuildInputs`. Did you try to use this package?

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix [hunk 8]
```diff
@@ -0,0 +1,27 @@
+{ lib, fetchurl, buildDunePackage, ocaml, js_of_ocaml, ppxlib, js_of_ocaml-ppx, gen_js_api }:
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/promise_jsoo/default.nix:1]:
```suggestion
{ lib, fetchurl, buildDunePackage, js_of_ocaml, ppxlib, js_of_ocaml-ppx, gen_js_api }:
```

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 2 above)
Comment (jayesh-bhoot) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix]: Yes. I am using that package here - https://github.com/jayesh-bhoot/pin-unpin-tab/blob/master/flake.nix

I am able to `dune build` and test the add-on.

The way I reason about this is:

`ojs` is a runtime dependency of `gen_js_api`. So, whichever package depends on `gen_js_api` should need `ojs` while building the package (i.e., `gen_js_api`'s runtime). So `ojs` should be available while building this package.

Now, X listed under `propagatedBuildInputs` in derivation A implies that if derivation B lists A under its `buildInputs`, then B will have X available automatically. So B would only have to specify A in its own `buildInputs`, and not explicitly keep track of X as well.

So, inside A.nix: `propagatedBuildInputs = [ X ];`.
Inside B.nix: `buildInputs = [ A ];`.
Now, X is available to B too.

With this in mind, `gen_js_api` derivation (https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/ocaml-modules/gen_js_api/default.nix) already lists `ojs` under its `propagatedBuildInputs`. So `promise_jsoo` which depends on `gen_js_api` need not depend explicitly on `ojs` too.

Of course, I am still learning all this. Let me know if I am wrong.

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 8 above)
Comment (jayesh-bhoot) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix:1]: Can you help me understand this suggestion?

Comment (vbgl) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix:1]: The derivation does not use the `ocaml` parameter. So there is no need to have it as a parameter.

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 2 above)
Comment (vbgl) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix]: This is right. But users of `promise_jsoo` may not depend on `gen_js_api`; nonetheless, they do depend on `ojs`.

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 8 above)
Comment (jayesh-bhoot) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix:1]: Oh crap. I misread the diff as you adding the `ocaml` parameter.

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 2 above)
Comment (jayesh-bhoot) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix]: Understood. `promise_jsoo` should work standalone.

File: pkgs/development/ocaml-modules/promise_jsoo/default.nix (hunk 8 above)
Comment (vbgl) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix]: It seems that `ocaml` is back into the parameter set.

Comment (jayesh-bhoot) [reply pkgs/development/ocaml-modules/promise_jsoo/default.nix]: Done.

**Pull request merged**